DOCS-982: Document `obstacles_depth` model of segmenter in vision service (#1721)
viamrobotics/docs: docs/services/vision/segmentation.md

- [**Obstacles point cloud (`obstacles_pointcloud`)**](#configure-an-obstacles_pointcloud-segmenter): A segmenter that identifies well-separated objects above a flat plane.
- [**Object detector (`detector_3d_segmenter`)**](#configure-a-detector_3d_segmenter): This model takes 2D bounding boxes from an object detector and projects the pixels in the bounding box to points in 3D space.

> - [**Obstacles depth (`obstacles_depth`)**](#configure-an-obstacles_depth-segmenter): A segmenter for depth cameras that returns a projection of a bounding box or boxes with a Pose as a vector from the depth camera lens to the center of the obstacle identified.

Reviewer: Can we change to "... that returns the obstacles as 3-dimensional boxes with a Pose as a vector.... "

Author: Done, perfect thank you!

Reviewer: Wait I'm sorry I messed up. I'm actually wrong because in some cases (no intrinsics) this isn't true. I think the most accurate thing is "A segmenter for depth cameras that returns the perceived obstacles as a set of one or more GeometriesInFrame, each with a Pose as a vector from the depth camera lens to the center of the obstacle identified."

Reviewer: This comment @sguequierre

> - [**Obstacles depth (`obstacles_depth`)**](#configure-an-obstacles_depth-segmenter): A segmenter for depth cameras that returns a projection of a bounding box or boxes with a Pose as a vector from the depth camera lens to the center of the obstacle identified.

Author: ```suggestion
- [**Obstacles depth (`obstacles_depth`)**](#configure-an-obstacles_depth-segmenter): A segmenter for depth cameras that returns the perceived obstacles as a set of 3-dimensional bounding boxes, each with a Pose as a vector.
```

Click **Save config** and proceed to [test your segmenter](#test-your-segmenter).

## Configure an `obstacles_depth` segmenter

This segmenter model is for depth cameras, and is best for motion planning with transient obstacles.

Use the segmenter to identify well separated objects above a flat plane.

> If you want to identify point clouds over the flat plane with your segmenter, [configure your frame system](/services/frame-system/#configuration) to configure the relative spatial orientation of the components of your robot within Viam's [frame system service](/services/frame-system/).

Author: ```suggestion
If you want to identify point clouds over the flat plane with your segmenter, [configure your frame system](/services/frame-system/#configuration) to configure the relative spatial orientation of the components of your robot, including your [camera](/components/camera/) within Viam's [frame system service](/services/frame-system/).
```

> If you want to identify point clouds over the flat plane with your segmenter, [configure your frame system](/services/frame-system/#configuration) to configure the relative spatial orientation of the components of your robot, including your [camera](/components/camera/) within Viam's [frame system service](/services/frame-system/).

Author: ```suggestion
If you want to identify point clouds over the flat plane with your segmenter, [configure your frame system](/services/frame-system/#configuration) to configure the relative spatial orientation of the components of your robot, including your [camera](/components/camera/), within Viam's [frame system service](/services/frame-system/).
```

> After configuring your frame system, your camera will populate its own `Properties` with these spatial intrinsic parameters from the frame system, as long you have configured the parameters as part of your camera.

Author: ```suggestion
After configuring your frame system, your camera will populate its own `Properties` with these spatial intrinsic parameters from the frame system.
```

You can then get those parameters from your camera through the [camera API](/components/camera/#getproperties).
Then, set the segmenter's attribute `with_geometries: true`.
The segmenter now returns point clouds within the `GeometryInFrame` object it captures.

> If you choose not to configure the frame system, you can still identify points over the flat plane with your segmenter.

Author: ```suggestion
If you choose not to configure the frame system, you can still identify points over the flat plane with your segmenter.
```
@npentrel Tried to rephrase this section to make it more clear what they are doing, which is getting the representation in one form or another. The attribute is just a safety check to determine what kind of representation you're doing instead of checking whether or not you've configured the frame system behind the scenes and deciding what representation from there-- AFAIK it could be done either way but this just allows the user to get points or point clouds if they want to, making it their choice. But if they want point clouds they must do frame system config.

Click **Save config** and proceed to [test your segmenter](#test-your-segmenter).

## Configure an `obstacles_depth` segmenter

This segmenter model is for depth cameras, and is best for motion planning with transient obstacles.

Use the segmenter to identify well separated objects above a flat plane.
If you want to identify well separated point clouds over the flat plane, set `with_geometries: true`.
[Configure your frame system](/services/frame-system/#configuration) to configure the relative spatial orientation of the components of your robot within Viam's [frame system service](/services/frame-system/).
After configuring your frame system, your camera will populate its own `Properties` with the intrinsic parameters from the frame system, as long you have configured the parameters as part of your camera.
You can then get those parameters from your camera through the [camera API](/components/camera/#getproperties).
The segmenter now returns point clouds within the `GeometryInFrame` object it captures.

> If you choose not to configure the frame system to provide instrinsic parameters, set `with_geometries: false`.

Author: @npentrel Added your suggestions here and above, I think! Let me know if it still doesn't make sense and we can maybe sync up next week to get this right?

Reviewer 2: I think I'm still not clear what I am identifying when `with_geometries: false`. On lines 208/209, we say the segmenter identifies well separated objects above a flat plane, and if we want then we should set `with_geometries: true`. But am I still identifying well separated objects when I have that set to `false`? Is the difference just that I don't get the coordinates of where those are? 

As a reader I still don't fully understand why I'd choose one over the other. An example may help here.

Click **Save config** and proceed to [test your segmenter](#test-your-segmenter).

## Configure an `obstacles_depth` segmenter

This segmenter model is for depth cameras, and is best for motion planning with transient obstacles.

> If you have [configured the frame system](/services/frame-system/#configuration) for the components of your robot, then you can implement this model to identify well separated objects above a flat plane.

Author: ```suggestion
After [configuring the frame system](/services/frame-system/#configuration) for the components of your robot, then you can implement this model to identify well separated objects above a flat plane.
```
@npentrel might be better

Reviewer 2: Ok I'm still confused about the ordering of decisions a reader has to make. I think this sentence should merge with the next paragraph to explain why you'd choose one or the other

Use the segmenter to identify well separated objects above a flat plane.

> Configure the `with_geometries` attribute according to whether you've [configured the frame system](/services/frame-system/#configuration) to provide intrinsic parameters:

Author: @npentrel Edited this re. your review/suggestion to combine diff sections

Reviewer 2: I still don't get this - you're saying here "wheter you've configured the frame system" and on 232 "start by configuring your frame system". Either you have already configured it or you start configuring it in 232. Or do you just configure part of the frame system subsequently. Also how do I choose whether I want to use a frame system. I think this should say something like "If you need to receive a point cloud, ..." "If you only need a single point, then you do not need ..."

> Configure the `with_geometries` attribute according to whether you've [configured the frame system](/services/frame-system/#configuration) to provide intrinsic parameters:

Author: ```suggestion
First, choose whether you want t configure the `with_geometries` attribute to provide intrinsic parameters from [configuring the frame system](/services/frame-system/#configuration).

The options are as follows:
```
I think I see why you're confused-- removed so it's not directing to configure-- what i mean to say is that you have to choose whether you want to or not as the beginning of the configuration process, bc this attribute should be determined  based on what you have done in the past -- configured the frame system or not. Does that make any sense? 

I would just remove this full section, but I think it's important to have the options listed so people can see immediately that their choice to configure frame system or not/selection for this attribute will then have this effect on what your segmenter returns

> To configure an `obstacles_depth` segmenter, first decide if you want to use the segmenter with or without intrinsic parameters and configure the `with_geometries` attribute accordingly:

Author: ```suggestion
To configure an `obstacles_depth` segmenter, first decide if you want to use the segmenter with or without frame system configuration.
Configure the `with_geometries` attribute accordingly:
```
@npentrel Suggestion re. your comment about intrinsic parameters: (1/2) (other one below)
I think this is mostly what is meant and might help to avoid confusion

Reviewer 2: oh so it's not intrinsic parameters? or are those the same?

> This segmenter model is for depth cameras, and is best for motion planning with transient obstacles.
> 
> Use the segmenter to identify well separated objects above a flat plane.
> Configure the `with_geometries` attribute according to whether you've [configured the frame system](/services/frame-system/#configuration) to provide intrinsic parameters:
> 
> - `true`: this segmenter will return point clouds within the `GeometryInFrame` object it captures.
> - `false`: this segmenter will return a single point within the `GeometryInFrame` object it captures.
> 
> If `true`, start by [configuring your frame system](/services/frame-system/#configuration) to configure the relative spatial orientation of the components of your robot within Viam's [frame system service](/services/frame-system/).

Reviewer 2: I think what this needs to do is explain better what the two options are. You're not choosing whether to configure a parameter to true or false, the reader should choose what functionality they need and based on that set the parameter. Now I am not entirely certain I am capturing the functionality correctly but the way I'd structure it is like this:

```suggestion
This segmenter model is for depth cameras, and is best for motion planning with transient obstacles.

You can use the segmenter to identify well seprated objects above a flat plane or to identify a single object (?):

- To identify multiple well separated objects above a flat plane, set `with_geometries: true`.
   Ensure that you have configured your camera's [`intrinsic_parameters`](https://docs.viam.com/components/camera/calibrate/).
   Then, [configure your frame system](/services/frame-system/#configuration) to configure the relative spatial orientation of the components of your robot within Viam's [frame system service](/services/frame-system/).
   After configuring your frame system, your camera populates its own `Properties` with the intrinsic parameters from the frame system (camera? the next line seems to indicate that?). 
  You can then get those parameters from your camera through the [camera API](/components/camera/#getproperties).
   The segmenter now returns point clouds within the `GeometryInFrame` object it captures.

- To identify a single object above a flat plane, set `with_geometries: false`.
   The segmenter now returns a single point within the `GeometryInFrame` object it captures.
```

I think this way it provides one choice point and only one mention of the frame system. Though I am not sure if the second one also needed the frame system - I guess that is also currently confusing: do I need to configure the frame system in both scenarios or just one?

Author: You don't need to/shouldn't have configured the frame system if `false`, but if you do and still set to false, that's fine, you just will get a point back anyways! Is that clear enough now? The parameter determines the behavior of the segmenter-- if you set to `true` and didn't configure anything I assume it will error out

> After configuring your frame system, your camera will populate its own `Properties` with the intrinsic parameters, as long you have configured the parameters as part of your camera.

Author: ```suggestion
After configuring your frame system, your camera will populate its own `Properties` with the intrinsic parameters from the frame system, as long you have configured the parameters as part of your camera.
```
@npentrel (2/2)

Click **Save config** and proceed to [test your segmenter](#test-your-segmenter).

## Configure an `obstacles_depth` segmenter

This segmenter model is for depth cameras, and is best for motion planning with transient obstacles.
If you have [configured the frame system](/services/frame-system/#configuration) for the components of your robot, then you can implement this model to identify well separated objects above a flat plane.

> To configure an `obstacles_depth` segmenter, first decide if you want to use the segmenter with or without intrinsic parameters.

Reviewer 2: Do we have anywhere we could link them to help understand what intrinsic parameters are? Maybe the camera page?

Author: Tried to address with 2 suggestioins below-- lmk what you think! @npentrel

> To configure an `obstacles_depth` segmenter, first decide if you want to use the segmenter with or without intrinsic parameters.

Author: ```suggestion
To configure an `obstacles_depth` segmenter, first decide if you want to use the segmenter with or without intrinsic parameters, attributes you configure to tell `viam-server` the physical properties of the robot.
```
@npentrel does this seem ok? I don't know what link would help explain either-- could be like "like this" and link to camera page if you think so

> This is determined by the value of the configuration attribute `return_pcds`:

Reviewer: Not true.. this is actually determined by **with_geometries**. 

return_pcds is a separate optional parameter (default false) that determines whether pointclouds should be included within the GeometryInFrame object. 

So if w_g and r_p were both true, you would get 10 GeometryInFrame objects, each with a geometry and a pointcloud inside. If w_g was false but r_p was true, you'd get a single GeometryInFrame object that would have a 1-point-pointcloud and a geometry (point).  When r_p is false, you get the same things just without the pointcloud. Geometry only.

Author: Ok tried to update!! Sorry for the confusion

> This is determined by the value of the configuration attribute `return_pcds`:

Author: ```suggestion
This is determined by the value of the configuration attribute `with_geometries`:
```

> This is determined by the value of the configuration attribute `with_geometries`:

Author: ```suggestion
This is determined by the value of the configuration attribute `return_pcds`:
```

> To configure an `obstacles_depth` segmenter, first decide if you want to use the segmenter with or without intrinsic parameters.
> 
> This is determined by the value of the configuration attribute `with_geometries`:

Reviewer 2: ```suggestion
To configure an `obstacles_depth` segmenter, first decide if you want to use the segmenter with or without intrinsic parameters and configure the `with_geometries` attribute accordingly:
```
I think it's unclear what "This" is and I don't like the passive voice. Is this still correct?

Author: Still correct, thank you for the suggestion and cleaning up the passive voice! Done

- `true`: this segmenter will return point clouds within the `GeometryInFrame` object it captures.
- `false`: this segmenter will return a single point within the `GeometryInFrame` object it captures.

> If `true`, start by [configuring your frame system](/services/frame-system/#configuration) to configure the relative spatial orientation of the components of your robot within Viam's [frame system service](/services/frame-system/).

Reviewer 2: ```suggestion
If `true`, start by [configuring your frame system](/services/frame-system/#configuration) to configure the relative spatial orientation of the components of your robot within Viam's [frame system service](/services/frame-system/).
```

Wait but in line 225 you imply it should already be configured

Author: I didn't mean to imply that it should already be configured in line 225, just that people can use it functionally this way if they do that. Does suggestion help?

> If `true`, start configuring your segmenter by following [these instructions](/services/frame-system/#configuration) to configure the relative spatial orientation of the components of your robot within Viam's [frame system service](/services/frame-system/).

Reviewer 2: ```suggestion
If `true`, start by [configuring your frame system](/services/frame-system/#configuration) to configure the relative spatial orientation of the components of your robot within Viam's [frame system service](/services/frame-system/).
```

Author: lgtm done!

> After doing this, your camera will populate its own `Properties` with the intrinsic parameters of the geometric configuration.

Reviewer 2: ```suggestion
After configuring your frame system, your camera will populate its own `Properties` with the intrinsic parameters of the geometric configuration.
```
 is this also only if the value is true? Maybe this should be in one paragraph then

> You can then get those parameters from your camera through the [camera API](/components/camera/#getproperties).

Author: Do you think I should remove then? I figured this could be helpful as a usage tip re. your email thread, but if not, can remove to help confusion

> You can then get those parameters from your camera through the [camera API](/components/camera/#getproperties).

Reviewer: Nah, the rewording as a usage tip is pretty nice I would just take away "of the geometric configuration." and replace it with "as long as they [the intrinsic parameters] have been configured as part of the camera."

> You can then get those parameters from your camera through the [camera API](/components/camera/#getproperties).

Author: Ok awesome done!

> You can then get those parameters from your camera through the [camera API](/components/camera/#getproperties).
> 
> Then, configure an `obstacles_depth` segmenter:

Reviewer 2: ```suggestion
To configure your segmented you need the Properties parameters from your camera which you can get with the [camera API](/components/camera/#getproperties).

Then, configure your `obstacles_depth` segmenter:
```

a -> your : because at this point we've given so much info it feels more specific

Author: Makes sense but not sure about needing properties here-- @kharijarrett if I implement this suggestion do you think line 237 makes sense/should be included?

{{< tabs >}}
{{% tab name="JSON Template" %}}

> Add the vision service object to the services array in your raw JSON configuration:

Reviewer 2: ```suggestion
Add the following vision service object to the services array in your raw JSON configuration:
```
 (to be more specific)

Author: Done

``` json {class="line-numbers linkable-line-numbers"}
"services": [
    {
    "name": "<segmenter_name>",
    "type": "vision",
    "model": "obstacles_pointcloud"
    "attributes": {
        "h_min_m": <number>,
        "h_max_m": <number>,
        "theta_max_deg": <number>,
        "return_pcds": <boolean>,
        "with_geometries": <boolean>,
    }
    },
    ... // Other services
]
```

{{% /tab %}}
{{% tab name="JSON Example" %}}

```json {class="line-numbers linkable-line-numbers"}
"services": [
{
    "name": "rc_segmenter",
    "type": "vision",
    "model": "obstacles_pointcloud",
    "attributes": {
        "h_min_m": 0.0,
        "h_max_m": 1.0,
        "theta_max_deg": 45,
        "return_pcds": "true",
        "with_geometries": "true"
    }
}
]
```

{{% /tab %}}
{{< /tabs >}}

> The following parameters are available for a `"obstacles_depth"` segmenter:

Reviewer 2: ```suggestion
The following parameters are available for an `"obstacles_depth"` segmenter:
```

Author: Done, thank you for catching

| Parameter | Inclusion | Description |
| --------- | --------- | ----------- |

> | `return_pcds` | Required | If you would like point clouds or a single point returned within their `GeometryInFrame` object. `true` indicates wanting point clouds. <br> Example: `false` </br> |

Author: ```suggestion
| `return_pcds` | Required | If you would like point clouds or a single point returned within the `GeometryInFrame` object captured by this segmenter. `true` indicates wanting point clouds. <br> Example: `false` </br> |
```

> | `return_pcds` | Required | Whether you would like point clouds or a single point returned within the `GeometryInFrame` object captured by this segmenter. If `true`, return point clouds. <br> Example: `"false"` </br> |

Author: ```suggestion
| `with_geometries` | Required | Whether you would like point clouds or a single point returned within the `GeometryInFrame` object captured by this segmenter. If `true`, return point clouds. <br> Example: `"false"` </br> |
```

> | `return_pcds` | Required | If you would like point clouds or a single point returned within the `GeometryInFrame` object captured by this segmenter. `true` indicates wanting point clouds. <br> Example: `"false"` </br> |

Reviewer 2: ```suggestion
| `return_pcds` | Required | Whether you would like point clouds or a single point returned within the `GeometryInFrame` object captured by this segmenter. `true` return point clouds. <br> Example: `"false"` </br> |
```

Author: Done edited your suggestion slight `true: return point clouds` to `If `true`, return point clouds`

> | `with_geometries` | Required | Whether you would like multiple boxes, if `true`, or a single point , if`false`, returned within the GeometryInFrame object captured by this segmenter.  <br> Example: `"false"` </br> |

Author: ```suggestion
| `with_geometries` | Required | Whether you would like multiple boxes, if `true`, or a single point, if `false`, returned within the `GeometryInFrame` object captured by this segmenter.  <br> Example: `"false"` </br> |
```

> | `with_geometries` | Required | Whether you would like point clouds or a single point returned within the `GeometryInFrame` object captured by this segmenter. If `true`, return point clouds. <br> Example: `"false"` </br> |

Reviewer: "Whether you would like multiple boxes (true) or a single point (false) returned within the GeometryInFrame object captured by this segmenter."

> | `h_min_m` | Optional | The minumum vertical height in meters acceptable so that any obstacle strictly lower than is not deemed an obstacle by the segmenter. <br> Default: `0.0` </br> |

Author: ```suggestion
| `h_min_m` | Optional | The minimum vertical height in meters acceptable so that any obstacle strictly lower than is not deemed an obstacle by the segmenter. <br> Default: `0.0` </br> |
```

> | `h_min_m` | Optional | The minimum vertical height in meters acceptable so that any obstacle strictly lower than is not deemed an obstacle by the segmenter. <br> Default: `0.0` </br> |

Reviewer 2: ```suggestion
| `h_min_m` | Optional | The minimum vertical height in meters for an object to be considered an obstacle. <br> Default: `0.0` </br> |
```

Author: done lgtm

> | `h_max_m` | Optional | The maximum vertical height in meters acceptable so that any obstacle strictly higher than is not deemed an obstacle by the segmenter. <br> Default: `1.0` </br> |

Reviewer 2: ```suggestion
| `h_max_m` | Optional | The maximum vertical height in meters at which an object is considered an obstacle. <br> Default: `1.0` </br> |
```

Author: Done

> | `theta_max_deg` | Optional | The largest slope in degrees acceptable for non-obstacle surface ramps. <br> Default: `45` </br> |

Author: ```suggestion
| `theta_max_deg` | Optional | The maximum slope at which an object can be considered an obstacle. <br> Default: `45` </br> |
```

Author: @kharijarrett Would this suggestion above be accurate? @npentrel Trying to match your suggested wording on other attributes

Reviewer: Not really because then it's the _minimum_ slope at which an object can be considered an obstacle. I think it makes sense to say "the maximum slope at which an object is still not an obstacle"

> | `theta_max_deg` | Optional | The largest slope in degrees acceptable for non-obstacle surface ramps. <br> Default: `45` </br> |

Reviewer 2: ```suggestion
| `theta_max_deg` | Optional | The largest slope in degrees acceptable for non-obstacle surface ramps. <br> Default: `45` </br> |
```
What does acceptable mean?

Author: Ok, edited w khari suggestion!

> | `return_pcds` | Optional | If you have configured the geometries or spatial orientation of the components of your robot, within the frame system.  <br> Example: `"false"` </br> |

Author: ```suggestion
| `return_pcds` | Optional | Whether you would like pointclouds to be included within the GeometryInFrame object captured by this segmenter.  <br> Example: `"false"` </br> |
```

> | `return_pcds` | Optional | If you have configured the geometries or spatial orientation of the components of your robot, within the frame system.  <br> Example: `"false"` </br> |

Author: @kharijarrett I flipped `with_geometries` and `return_pcds` re. your feedback-- is this accurate now or still flipped up?

Reviewer: "Whether you would like pointclouds to be included within the GeometryInFrame object captured by this segmenter."

Author: Thank you, done/added!

> | `with_geometries` | Optional | If you have configured the geometries or spatial orientation of the components of your robot, within the frame system. `true` is necessary to `return_pcds` properly, if `"return_pcds": "true"`. <br> Example: `"false"` </br> |

Author: ```suggestion
| `return_pcds` | Optional | If you have configured the geometries or spatial orientation of the components of your robot, within the frame system.  <br> Example: `"false"` </br> |
```

> | `with_geometries` | Optional | If you have configured the geometries, or, spatial orientation of the components of your robot, within the frame system. `true` is necessary to `return_pcds` properly, if `"return_pcds": "true"`  <br> Example: `false` </br> |

Author: ```suggestion
| `with_geometries` | Optional | If you have configured the geometries, or, spatial orientation of the components of your robot, within the frame system. `true` is necessary to `return_pcds` properly, if `"return_pcds": "true"`  <br> Example: `"false"` </br> |
```

> | `with_geometries` | Optional | If you have configured the geometries, or, spatial orientation of the components of your robot, within the frame system. `true` is necessary to `return_pcds` properly, if `"return_pcds": "true"`  <br> Example: `false` </br> |

Author: @kharijarrett Any suggestions here on the tables are very welcome, I'm not sure if I'm phrasing properly esp. `with_geometries`

> | `with_geometries` | Optional | If you have configured the geometries, or, spatial orientation of the components of your robot, within the frame system. `true` is necessary to `return_pcds` properly, if `"return_pcds": "true"`  <br> Example: `"false"` </br> |

Reviewer 2: ```suggestion
| `with_geometries` | Optional | If you have configured the geometries or spatial orientation of the components of your robot, within the frame system. `true` is necessary to `return_pcds` properly, if `"return_pcds": "true"`. <br> Example: `"false"` </br> |
```

Author: Done
